src/ml_models.py

```diff
@@ -334,24 +339,56 @@ def _create_optimizer(self) -> optim.Optimizer:
         """Create an optimizer after the model is initialized."""
         return optim.Adam(self.parameters(), lr=self._learning_rate)
 
+    def _reset_weights(self) -> None:
+        """(Re-)initialize the network weights."""
+        self.apply(lambda m: self._weight_reset(m, self._weight_init, self.
+                                                _weight_init_param))
+
+    @staticmethod
+    def _weight_reset(m: torch.nn.Module, weight_init: str,
+                      param: Optional[float]) -> None:
+        if isinstance(m, nn.Linear):
+            if weight_init == "default":
+                m.reset_parameters()
+            elif weight_init == "uniform":
+                assert param is not None
+                torch.nn.init.uniform_(m.weight, -param, param)
+            elif weight_init == "normal":
+                assert param is not None
+                torch.nn.init.normal_(m.weight, std=param)
+            else:
+                raise NotImplementedError(
+                    f"{weight_init} weight initialization"
+                    " unknown")
+        raise NotImplementedError(f"Module type {type(m)} not supported")
+
     def _fit(self, X: Array, y: Array) -> None:
         # Initialize the network.
         self._initialize_net()
-        # Create the loss function.
-        loss_fn = self._create_loss_fn()
-        # Create the optimizer.
-        optimizer = self._create_optimizer()
         # Convert data to tensors.
         tensor_X = torch.from_numpy(np.array(X, dtype=np.float32))
         tensor_y = torch.from_numpy(np.array(y, dtype=np.float32))
         batch_generator = _single_batch_generator(tensor_X, tensor_y)
-        # Run training.
-        _train_pytorch_model(self,
-                             loss_fn,
-                             optimizer,
-                             batch_generator,
-                             max_iters=self._max_train_iters,
-                             n_iter_no_change=self._n_iter_no_change)
+        for _ in range(self._n_reinitialize_tries):
+            # (Re-)initialize weights.
+            self._reset_weights()
+            # Create the loss function.
+            loss_fn = self._create_loss_fn()
+            # Create the optimizer.
+            optimizer = self._create_optimizer()
```

Comment:
same question about the optimizer, although _maybe_ it has some kind of internal state